### ISSUE TITLE
NDArray backend does not converge, even though WGPU backend does

### ISSUE THREAD

**Issue by jsosulski** [tracel-ai/burn#3565]
**Describe the bug**

I followed the burn book to reproduce the MNIST training example. Out of interest, I wanted to try out the `NDArray` backend, but the network does not converge -> it has much worse validation accuracy with ndarray backend.

**To Reproduce**

This repo contains the mnist code from the book:

https://github.com/jsosulski/mnist_mwe

To run on cpu: `cargo run -r` and to  run on gpu use `cargo run -r --features wgpu`.

**Expected behavior**

Regardless of backend choice, the network reaches at least somewhat similar validation accuracies.

**Screenshots**

This is the output for the training on WGPU backend:

```bash
$ cargo run --release --features wgpu
...
| Split | Metric   | Min.     | Epoch    | Max.     | Epoch    |
|-------|----------|----------|----------|----------|----------|
| Train | Accuracy | 82.317   | 1        | 82.317   | 1        |
| Train | Loss     | 0.615    | 1        | 0.615    | 1        |
| Valid | Accuracy | 92.300   | 1        | 92.300   | 1        |
| Valid | Loss     | 0.252    | 1        | 0.252    | 1        |
```

This is the output for the training on NDArray backend:

```bash
$ cargo run --release
...
| Split | Metric   | Min.     | Epoch    | Max.     | Epoch    |
|-------|----------|----------|----------|----------|----------|
| Train | Accuracy | 73.732   | 1        | 73.732   | 1        |
| Train | Loss     | 1.006    | 1        | 1.006    | 1        |
| Valid | Accuracy | 8.470    | 1        | 8.470    | 1        |
| Valid | Loss     | 3.637    | 1        | 3.637    | 1        |
```

**Desktop (please complete the following information):**
 - OS: Manjaro
 - Version: stable

**Additional context**

If it is relevant / helps, I am on a Laptop that does not have a dedicated GPU. The CPU is: `AMD Ryzen 7 PRO 5850U with Radeon Graphics (16) @ 4.507GHz`.
I *also* tried on latest main, but have the same results.

Probably offtopic, but on main I had to remove the `.devices(...)` call from the `learner` and as it turns out, this is also not needed on 0.18.0.

**Comment by laggui** [comment]
Sorry for the late reply.

Does it converge after more than one epoch? For a single epoch, the difference could be in the random initialization.

**Comment by jsosulski** [comment]
No worries!

I initially observed it with 3 epochs, but as 1 epoch already converges with WGPU, I used that for the MWE repo. But here is the output when using 10 epochs:

Using **WGPU**
```
Total Epochs: 10

| Split | Metric   | Min.     | Epoch    | Max.     | Epoch    |
|-------|----------|----------|----------|----------|----------|
| Train | Accuracy | 82.138   | 1        | 97.242   | 10       |
| Train | Loss     | 0.090    | 10       | 0.632    | 1        |
| Valid | Accuracy | 92.200   | 1        | 98.100   | 10       |
| Valid | Loss     | 0.053    | 10       | 0.252    | 1        |
```

Using **ndarray**
```
Total Epochs: 10

| Split | Metric   | Min.     | Epoch    | Max.     | Epoch    |
|-------|----------|----------|----------|----------|----------|
| Train | Accuracy | 69.653   | 1        | 86.425   | 10       |
| Train | Loss     | 0.428    | 10       | 1.154    | 1        |
| Valid | Accuracy | 5.780    | 6        | 14.110   | 2        |
| Valid | Loss     | 2.957    | 1        | 8.045    | 10       |
```

While 14% is better than 8%, in a 10-class problem it is IMO still fairly close to chance level. I can try to let the training run for 200 or more epochs, but this will be a task for overnight. 10 epochs took ~20minutes. If 200 epochs converge, then there's still the question why ndarray converges much more slowly.

As someone who hasn't really worked with burn before, I see one key difference -> Kernel Fusion is not available for ndarray and for wgpu it is automatically enabled, afaict.

**Comment by laggui** [comment]
Thanks for checking with more epochs!

There does seem to be an issue, at 10 epochs it should at least be trending towards a converging solution. That is apparently not the case here with the reported ndarray results. No need to train longer, thanks for reporting back 🙂 

It should be easy to test without fusion for wgpu, the default features can be disabled. I doubt that will make wgpu *not* converge in 10 epochs 😅 so might be an incorrect implementation a precision issue in the ndarray backend implementation. Will have to take a closer look!

**Comment by jsosulski** [comment]
Yeah I disabled fusion and still WGPU results in a well-performing model even after 1 epoch. Let me know if I should try anything else!

**Comment by laggui** [comment]
Thanks for all the feedback 🙏 The example is relatively straightforward so I don't have anything else in mind, will probably have to check it out to see if it is a precision issue / wrong calculation that's causing this.

**Comment by antimora** [comment]
The ndarray conv implementation had a correctness fix since (#4727), which could explain the gap. flex is the default CPU backend now, so a retest with flex would also help. Could you retest on current main (or 0.22 once it's out)? If we don't hear back in a few weeks we'll assume it's fixed and close this.
